kaldb/src/test/java/com/slack/kaldb/clusterManager/SnapshotDeletionServiceTest.java

```diff
@@ -0,0 +1,798 @@
+            .build();
+
+    SnapshotDeletionService snapshotDeletionService =
+        new SnapshotDeletionService(
+            replicaMetadataStore, snapshotMetadataStore, s3BlobFs, managerConfig, meterRegistry);
+
+    int deletes = snapshotDeletionService.deleteExpiredSnapshotsWithoutReplicas();
+    assertThat(deletes).isEqualTo(0);
+
+    assertThat(snapshotMetadataStore.getCached().size()).isEqualTo(0);
+    assertThat(replicaMetadataStore.getCached().size()).isEqualTo(0);
+    verify(s3BlobFs, times(0)).delete(any(), anyBoolean());
+
+    assertThat(MetricsUtil.getCount(SnapshotDeletionService.SNAPSHOT_DELETE_SUCCESS, meterRegistry))
+        .isEqualTo(0);
+    assertThat(MetricsUtil.getCount(SnapshotDeletionService.SNAPSHOT_DELETE_FAILED, meterRegistry))
+        .isEqualTo(0);
+    assertThat(
+            MetricsUtil.getTimerCount(SnapshotDeletionService.SNAPSHOT_DELETE_TIMER, meterRegistry))
+        .isEqualTo(1);
+  }
+
+  @Test
+  public void shouldHandleNoReplicasUnexpiredSnapshots() throws Exception {
+    KaldbConfigs.ManagerConfig.ReplicaCreationServiceConfig replicaCreationServiceConfig =
+        KaldbConfigs.ManagerConfig.ReplicaCreationServiceConfig.newBuilder()
+            .setReplicaLifespanMins(1440)
+            .build();
+
+    KaldbConfigs.ManagerConfig.SnapshotDeletionServiceConfig snapshotDeletionServiceConfig =
+        KaldbConfigs.ManagerConfig.SnapshotDeletionServiceConfig.newBuilder()
+            .setSchedulePeriodMins(10)
+            .setSnapshotLifespanMins(10080)
+            .build();
+
+    KaldbConfigs.ManagerConfig managerConfig =
+        KaldbConfigs.ManagerConfig.newBuilder()
+            .setReplicaCreationServiceConfig(replicaCreationServiceConfig)
+            .setSnapshotDeletionServiceConfig(snapshotDeletionServiceConfig)
+            .setScheduleInitialDelayMins(0)
+            .build();
+
+    Path file = Files.createTempFile("", "");
+    URI path = createURI(S3_TEST_BUCKET, "foo", "bar");
+    s3BlobFs.copyFromLocalFile(file.toFile(), path);
+
+    SnapshotMetadata snapshotMetadata =
+        new SnapshotMetadata(
+            UUID.randomUUID().toString(),
+            path.toString(),
+            Instant.now().minus(11000, ChronoUnit.MINUTES).toEpochMilli(),
+            Instant.now().minus(500, ChronoUnit.MINUTES).toEpochMilli(),
+            0,
+            "1");
+    snapshotMetadataStore.create(snapshotMetadata);
+    await().until(() -> snapshotMetadataStore.getCached().size() == 1);
+
+    SnapshotDeletionService snapshotDeletionService =
+        new SnapshotDeletionService(
+            replicaMetadataStore, snapshotMetadataStore, s3BlobFs, managerConfig, meterRegistry);
+
+    int deletes = snapshotDeletionService.deleteExpiredSnapshotsWithoutReplicas();
+    assertThat(deletes).isEqualTo(0);
+
+    assertThat(snapshotMetadataStore.getCached()).containsExactlyInAnyOrder(snapshotMetadata);
+    verify(s3BlobFs, times(0)).delete(any(), anyBoolean());
```

Comment:
Please add an additional check to ensure that the s3 path and the data exists?

Reply:
Done. Added additional checks for the s3BlobFs for every test that attempts to invoke delete with existing files. This shouldn't be strictly necessary since we verified `delete()` wasn't called, but in the event something were to change and an overloaded `delete()` implementation were to be called this should double check the file integrity.